toolchains/node/node_toolchain.bzl

```diff
@@ -0,0 +1,50 @@
+# Copyright 2018 The Bazel Authors. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+This module implements the node toolchain rule.
+"""
+
+NodeInfo = provider(
+    doc = "Information about how to invoke the node binary.",
+    fields = {
+        "target_tool": "A hermetically downloaded nodejs executable target.",
+        "target_tool_path": "Path to an existing nodejs executable",
+    },
+)
+
+def _node_toolchain_impl(ctx):
+    if not ctx.attr.target_tool and not ctx.attr.target_tool_path:
+        print("No nodejs binary was found or built, executing run for rules_nodejs targets might not work.")
```

Comment:
Should this be a fail? I can't think of a case where this would be desirable.

Reply:
Further below we are falling back to looking for `node` on the PATH if nothing was provided. Again that approach is taken from rules_k8s. But maybe we do not want that?

Reply:
Ahhh. I see. I node_toolchain_configure you look for "node":
```
def _impl(repository_ctx):
    if repository_ctx.attr.target_tool and repository_ctx.attr.target_tool_path:
        fail("Can only set one of target_tool or target_tool_path but both where set.")

    if repository_ctx.attr.target_tool:
        substitutions = {"%{TOOL_ATTRS}": "    target_tool = \"%s\"\n" % repository_ctx.attr.target_tool}
    else:
        default_tool_path = repository_ctx.attr.target_tool_path or repository_ctx.which("node") or ""
        substitutions = {"%{TOOL_ATTRS}": "    target_tool_path = \"%s\"\n" % default_tool_path}
```
but then it would be set at this point if it was found... and if not then it found fail further down at this point:
```
    if node_tool_info.target_tool_path == "" and not node_tool_info.target_tool:
        # If tool_path is empty and tool_target is None then there is no local
        # node tool, we will just print a nice error message if the user
        # attempts to do bazel run
        fail("The node toolchain was not properly configured so %s cannot be executed. Make sure that target_tool_path or target_tool is set." % ctx.attr.name)
```

Reply:
Cleaned this up